src/ws/cockpitauthpipe.c

```diff
@@ -0,0 +1,466 @@
+  g_object_class_install_property (gobject_class, PROP_PIPE_TIMEOUT,
+                                   g_param_spec_uint ("pipe-timeout",
+                                                        NULL,
+                                                        NULL,
+                                                        0,
+                                                        900,
+                                                        30,
+                                                        G_PARAM_WRITABLE |
+                                                        G_PARAM_CONSTRUCT_ONLY |
+                                                        G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_ID,
+                                   g_param_spec_string ("id",
+                                                        NULL,
+                                                        NULL,
+                                                        NULL,
+                                                        G_PARAM_WRITABLE |
+                                                        G_PARAM_CONSTRUCT_ONLY |
+                                                        G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_LOGNAME,
+                                   g_param_spec_string ("logname",
+                                                        NULL,
+                                                        NULL,
+                                                        NULL,
+                                                        G_PARAM_WRITABLE |
+                                                        G_PARAM_CONSTRUCT_ONLY |
+                                                        G_PARAM_STATIC_STRINGS));
+
+  signals[CLOSED] = g_signal_new ("close", COCKPIT_TYPE_AUTH_PIPE, G_SIGNAL_RUN_FIRST,
+                                  0, NULL, NULL, g_cclosure_marshal_generic,
+                                  G_TYPE_NONE, 1, G_TYPE_ERROR);
+
+  signals[MESSAGE] = g_signal_new ("message", COCKPIT_TYPE_AUTH_PIPE, G_SIGNAL_RUN_FIRST,
+                                   0, NULL, NULL, g_cclosure_marshal_generic,
+                                   G_TYPE_NONE, 1, G_TYPE_BYTES);
+
+}
+
+void
+cockpit_auth_pipe_close (CockpitAuthPipe *self,
+                         const gchar *problem)
+{
+  if (self->closed)
+    return;
+
+  self->closed = TRUE;
+
+  if (self->timeout)
+    g_source_remove (self->timeout);
+  self->timeout = 0;
+
+  if (self->sig_pipe_read > 0)
+    g_signal_handler_disconnect (self->pipe, self->sig_pipe_read);
+  self->sig_pipe_read = 0;
+
+  if (!self->pipe_closed)
+    cockpit_pipe_close (self->pipe, problem);
+
+  if (self->sig_pipe_close > 0)
+    g_signal_handler_disconnect (self->pipe, self->sig_pipe_close);
+  self->sig_pipe_close = 0;
+
+  g_clear_object (&self->pipe);
+}
+
+int
+cockpit_auth_pipe_claim_fd (CockpitAuthPipe *self)
```

Comment:
Yeah, lets change it for ease of readability. This stuff is complex enough as it is.